### PR TITLE
placement: add rule definitions

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve? 
Add definitions for placement rules feature.

### What is changed and how it works?
cannot work now.

### Check List 

Tests 
 - Unit test

### CODE REVIEW

File: server/schedule/placement/label_constraint.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+var exclusiveLabels = []string{"engine"}
+
+// MatchLabelConstraints checks if a store matches label constraints list.
+func MatchLabelConstraints(store *core.StoreInfo, constraints []LabelConstraint) bool {
+	return store != nil &&
+		slice.AllOf(constraints, func(i int) bool { return constraints[i].MatchStore(store) }) &&
+		slice.NoneOf(exclusiveLabels, func(i int) bool {
+			label := exclusiveLabels[i]
+			return store.GetLabelValue(label) != "" &&
+				slice.NoneOf(constraints, func(i int) bool { return constraints[i].Key == label })
```
Comment (lhy1024) [inline server/schedule/placement/label_constraint.go]: This judgment involving double negation is complicated. Do we have a way to simplify it?

File: server/schedule/placement/rule.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// Rules are ordered by (groupID, Index, ID).
+func compareRule(a, b *Rule) int {
+	switch {
+	case a.GroupID < b.GroupID:
+		return -1
+	case a.GroupID > b.GroupID:
+		return 1
+	case a.Index < b.Index:
```
Comment (lhy1024) [inline server/schedule/placement/rule.go]: does it mean index order priority?

File: server/schedule/placement/label_constraint.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+// LabelConstraintOp defines how a LabelConstraint matches a store. It can be one of
+// 'in', 'notIn', 'exists', or 'notExists'.
+type LabelConstraintOp string
+
+const (
+	// In restricts the store label value should in the value list.
+	In LabelConstraintOp = "in"
+	// NotIn restricts the store label value should not in the value list.
```
Comment (nolouch) [inline server/schedule/placement/label_constraint.go]: Should specify the store should have the key? sometimes the target set may `NotIn`&&`NotExists`.

File: server/schedule/placement/rule.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	case a.ID > b.ID:
+		return 1
+	default:
+		return 0
+	}
+}
+
+func sortRules(rules []*Rule) []*Rule {
+	sort.Slice(rules, func(i, j int) bool { return compareRule(rules[i], rules[j]) < 0 })
+	return rules
```
Comment (rleungx) [inline server/schedule/placement/rule.go]: It seems no need to return value?

File: server/schedule/placement/label_constraint.go (hunk 1 above)
Comment (disksing) [reply server/schedule/placement/label_constraint.go]: good point.

File: server/schedule/placement/rule.go (hunk 2 above)
Comment (disksing) [reply server/schedule/placement/rule.go]: yes

File: server/schedule/placement/rule.go (hunk 4 above)
Comment (disksing) [reply server/schedule/placement/rule.go]: good point.

File: server/schedule/placement/label_constraint.go (hunk 3 above)
Comment (disksing) [reply server/schedule/placement/label_constraint.go]: I think it is reasonable that both `notIn` and `notExists` be satisfied when a store does not have the key.

File: server/schedule/placement/rule.go (hunk 2 above)
Comment (lhy1024) [reply server/schedule/placement/rule.go]: I think we can add a comment to make it clearer

File: server/schedule/placement/label_constraint.go (hunk 1 above)
Comment (disksing) [reply server/schedule/placement/label_constraint.go]: Please check if the new version better. 🤝

File: server/schedule/placement/rule.go (hunk 2 above)
Comment (disksing) [reply server/schedule/placement/rule.go]: Add some comments to `Rule`

File: server/schedule/placement/label_constraint.go (hunk 1 above)
Comment (lhy1024) [reply server/schedule/placement/label_constraint.go]: Good. It is very clear :100:

File: server/schedule/placement/label_constraint.go (hunk 3 above)
Comment (disksing) [reply server/schedule/placement/label_constraint.go]: fixed.

File: server/schedule/placement/label_constraint.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+// MatchLabelConstraints checks if a store matches label constraints list.
+func MatchLabelConstraints(store *core.StoreInfo, constraints []LabelConstraint) bool {
+	if store == nil {
+		return false
+	}
+
+	if slice.AnyOf(exclusiveLabels, func(i int) bool { // if there is any exclusive label that
+		label := exclusiveLabels[i]
+		return store.GetLabelValue(label) != "" && // ... the store has the label
```
Comment (lhy1024) [inline server/schedule/placement/label_constraint.go]:
```suggestion
		return store.GetLabelValue(label) != "" && // ... the store has the exclusive label
```

File: server/schedule/placement/label_constraint.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+			slice.NoneOf(constraints, func(i int) bool { return constraints[i].Key == label }) // ... but label is not in constraints
```
Comment (lhy1024) [inline server/schedule/placement/label_constraint.go]:
```suggestion
			slice.NoneOf(constraints, func(i int) bool { return constraints[i].Key == label }) // ... but the exclusive label is not in constraints
```

File: server/schedule/placement/label_constraint.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	// NotIn restricts the store label value should not in the value list.
+	// If label does not exist, `notIn` is always true.
+	NotIn LabelConstraintOp = "notIn"
+	// Exists restricts the store should have the label.
+	Exists LabelConstraintOp = "exists"
+	// NotExists restricts the store should not have the label.
+	NotExists LabelConstraintOp = "notExists"
+)
+
+func validateOp(op LabelConstraintOp) bool {
```
Comment (lhy1024) [inline server/schedule/placement/label_constraint.go]: It seem to be not used in this PR?

Comment (disksing) [reply server/schedule/placement/label_constraint.go]: Don't worry. Later PR will use it.

File: server/schedule/placement/rule.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+func validateRole(s PeerRoleType) bool {
+	return s == Voter || s == Leader || s == Follower || s == Learner
+}
+
+// Rule is the placement rule that can be checked against a region. When
+// applying rules (apply means schedule regions to match selected rules), the
+// apply order is defined by the tuple [GroupID, Index, ID].
+type Rule struct {
+	GroupID          string            `json:"group_id"`                    // mark the source that add the rule
```
Comment (rleungx) [inline server/schedule/placement/rule.go]: BTW, I think it's better to unify the style of it in the future. We have both `snake_case` and `kebab-case` in our repo.

Comment (disksing) [reply server/schedule/placement/rule.go]: em, so are you suggesting using kebab style?

Comment (disksing) [reply server/schedule/placement/rule.go]: Maybe we can decide later, now both styles are used commonly.

Comment (rleungx) [reply server/schedule/placement/rule.go]: Of course.

**Pull request merged**